pipeline/matching/core/analytics.py

```diff
@@ -0,0 +1,27 @@
+from pipeline.matching.core import MatchingEngine
+from pipeline.types import MatchingInput, MatchingInternalAnalytics
+
+
+class CollectInternalAnalytics:
+    def __init__(self, name, matching_engine):
+        self.name: str = name
+        self.matching_engine: MatchingEngine = matching_engine
+
+    def n_proposed_matches_per_user(
+        self, inp: MatchingInput
+    ) -> MatchingInternalAnalytics:
+
+        n_proposed_matches_per_user = []
+        for user in inp.users:
+            match_count = 0
+            for generator in self.matching_engine.generators:
+                gen_matches = generator.generate(inp)
+                for rank_match in self.matching_engine.ranker.rank(
+                    inp, gen_matches
+                ):
+                    if user.uid in rank_match.users:
+                        match_count += 1
```

Comment:
instead of doing the actual generation and ranking, this component should only be doing analytics.

So what we want is to take the generator of ranked matches as an input, and then include the analytics in the main engine.

Change the input of this method to take `Iterator[Match]` you can also have the `MatchingInput` as an input, but I don't think we need it for anything. Can you think of anything?

Then we can integrate your analytics class to run after line 41 of the core matching engine code. That way, the matches will already be generated and ranked, you can pass them through your analytics code to collect the counts, and then return the generator of matches untouched and the matching engine code can continue as if nothing happened.

Reply:
To be even more clear, this is how the analytics code should integrate into the core matching engine:

```python
    def run(self, inp: MatchingInput) -> MatchingOutput:
        # Generate matches
        proposed: Iterator[Match] = self.generate_matches(inp)
        # Rank matches
        ranked: Iterator[Match] = self.rank_matches(inp, proposed)
        #################
        # This part is where your code goes:
        analyzer = CollectInternalAnalytics
        collected: Iterator[Match] = analyzer.n_proposed_matches_per_user(ranked)
        # Don't forget to change the input on the line below from ranked to collected
        #################
        # Select priority matches and get remaining users
        results = self.select_priority_matches(inp, collected) # Change the input on this line too
        selected: List[Match] = results[0]
        users: Set[UserId] = results[1]
        # Finalize remaining matches
        finalized: List[Match] = self.finalize_matches(inp, selected, users)
```

The reason we want to do it this way is for separation of concerns:
- Analytics code should only be worried about analytics
- Analytics code should not actually run the generators or rankers, just take their output
- We should be able to make some kinds of changes to the logic of the matching engine and still use the analytics code
- Unit tests for analytics code do not need any generators or rankers, just an iterator of matches, they become much simpler to test
- We can also write an integration test that shows how the analyzer and engine work together, that will be a stretch goal, we can pair on that unless you already know how to write it

Reply:
I see. I was under the impression I wasn't meant to alter the existing matching engine code or its associated functions. But if this is the case, then I see where I went wrong and have a better idea of what I should do going forward. I will revise accordingly.

Reply:
Also don't we need MatchingInput as an input in order to pull all the users? I could be wrong but that was my assumption. I'll add MatchingInput as an input but will revise accordingly if it turns out to be unnecessary.

Reply:
After our pairing call, Chris and I realized that it would be much easier to decouple the analytics from the matching engine code, that way they can function separately. The interface between them is that we now return internal data from the matching engine that the analytics can use. Merged in #242.